controller/gke-cluster-config-handler.go

```diff
@@ -261,19 +272,87 @@ func (h *Handler) enqueueUpdate(config *gkev1.GKEClusterConfig) (*gkev1.GKEClust
 	return h.gkeCC.UpdateStatus(config)
 }
 
-func (h *Handler) updateUpstreamClusterState(config *gkev1.GKEClusterConfig, upstreamSpec *gkev1.GKEClusterConfigSpec, svc *gkeapi.Service) (*gkev1.GKEClusterConfig, error) {
+func (h *Handler) updateUpstreamClusterState(config *gkev1.GKEClusterConfig, upstreamSpec *gkev1.GKEClusterConfigSpec) (*gkev1.GKEClusterConfig, error) {
 
-	if config.Spec.KubernetesVersion != nil {
-		if utils.StringValue(upstreamSpec.KubernetesVersion) != utils.StringValue(config.Spec.KubernetesVersion) {
-			logrus.Infof("updating kubernetes version for cluster [%s]", config.Name)
-			err := utils.UpdateCluster(config, &gkeapi.UpdateClusterRequest{
-				Update: &gkeapi.ClusterUpdate{
-					DesiredMasterVersion: *config.Spec.KubernetesVersion,
-				}})
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	cred, err := getSecret(ctx, h.secretsCache, &config.Spec)
+	if err != nil {
+		return config, err
+	}
+
+	changed, err := gke.UpdateMasterKubernetesVersion(cred, config, upstreamSpec)
+	if err != nil {
+		return config, err
+	}
+	if changed == gke.Changed {
+		return h.enqueueUpdate(config)
+	}
+
+	changed, err = gke.UpdateClusterAddons(cred, config, upstreamSpec)
+	if err != nil {
+		return config, err
+	}
+	if changed == gke.Retry {
+		h.gkeEnqueueAfter(config.Namespace, config.Name, wait*time.Second)
+		return config, nil
+	}
+	if changed == gke.Changed {
+		return h.enqueueUpdate(config)
+	}
+
+	changed, err = gke.UpdateMasterAuthorizedNetworks(cred, config, upstreamSpec)
+	if err != nil {
+		return config, err
+	}
+	if changed == gke.Changed {
+		return h.enqueueUpdate(config)
+	}
+
+	changed, err = gke.UpdateLoggingMonitoringService(cred, config, upstreamSpec)
+	if err != nil {
+		return config, err
+	}
+	if changed == gke.Changed {
+		return h.enqueueUpdate(config)
+	}
+
+	changed, err = gke.UpdateNetworkPolicy(cred, config, upstreamSpec)
+	if err != nil {
+		return config, err
+	}
+	if changed == gke.Changed {
+		return h.enqueueUpdate(config)
+	}
+
+	if config.Spec.NodePools != nil {
```

Comment:
Adding and removing node pools should also be supported. Would that be added in another PR?